cudax/include/cuda/experimental/__container/async_mdarray.cuh

```diff
@@ -0,0 +1,901 @@
+  _CCCL_NODISCARD _CCCL_HIDE_FROM_ABI const_pointer __unwrapped_end() const noexcept
+  {
+    return __buf_.data() + size();
+  }
+#endif // _CCCL_DOXYGEN_INVOKED
+
+  //! @}
+
+  //! @addtogroup access
+  //! @{
+  //! @brief Returns a reference to the \p __n 'th element of the async_mdarray
+  //! @param __n The index of the element we want to access
+  _CCCL_NODISCARD _CCCL_HIDE_FROM_ABI reference operator[](const size_type __n) noexcept
+  {
+    _CCCL_ASSERT(__n < size(), "cuda::experimental::async_mdarray subscript out of range!");
+    return begin()[__n];
+  }
+
+  //! @brief Returns a reference to the \p __n 'th element of the async_mdarray
+  //! @param __n The index of the element we want to access
+  _CCCL_NODISCARD _CCCL_HIDE_FROM_ABI const_reference operator[](const size_type __n) const noexcept
+  {
+    _CCCL_ASSERT(__n < size(), "cuda::experimental::async_mdarray subscript out of range!");
+    return begin()[__n];
+  }
+
+  //! @brief Returns a reference to the first element of the async_mdarray
+  _CCCL_NODISCARD _CCCL_HIDE_FROM_ABI reference front() noexcept
+  {
+    _CCCL_ASSERT(size() != 0, "cuda::experimental::async_mdarray front() called on empty async_mdarray!");
+    return begin()[0];
+  }
+
+  //! @brief Returns a reference to the first element of the async_mdarray
+  _CCCL_NODISCARD _CCCL_HIDE_FROM_ABI const_reference front() const noexcept
+  {
+    _CCCL_ASSERT(size() != 0, "cuda::experimental::async_mdarray front() called on empty async_mdarray!");
+    return begin()[0];
+  }
+
+  //! @brief Returns a reference to the last element of the async_mdarray
+  _CCCL_NODISCARD _CCCL_HIDE_FROM_ABI reference back() noexcept
+  {
+    _CCCL_ASSERT(size() != 0, "cuda::experimental::async_mdarray back() called on empty async_mdarray!");
+    return begin()[size() - 1];
+  }
+
+  //! @brief Returns a reference to the last element of the async_mdarray
+  _CCCL_NODISCARD _CCCL_HIDE_FROM_ABI const_reference back() const noexcept
+  {
+    _CCCL_ASSERT(size() != 0, "cuda::experimental::async_mdarray back() called on empty async_mdarray!");
+    return begin()[size() - 1];
+  }
+  //! @}
+
+  //! @addtogroup capacity
+  //! @{
+  //! @brief Returns the current number of elements stored in the async_mdarray.
+  _CCCL_NODISCARD _CCCL_HIDE_FROM_ABI size_type size() const noexcept
+  {
+    return __size(__mapping_.extents());
+  }
+
+  //! @brief Returns true if the async_mdarray is empty.
+  _CCCL_NODISCARD _CCCL_HIDE_FROM_ABI bool empty() const noexcept
+  {
+    return size() == 0;
+  }
+
+  //! @brief Returns the capacity of the current allocation of the async_mdarray..
+  _CCCL_NODISCARD _CCCL_HIDE_FROM_ABI size_type capacity() const noexcept
+  {
+    return size();
+  }
+
+  //! @brief Returns the maximal size of the async_mdarray.
+  _CCCL_NODISCARD _CCCL_HIDE_FROM_ABI constexpr size_type max_size() const noexcept
+  {
+    return static_cast<size_type>((_CUDA_VSTD::numeric_limits<difference_type>::max)());
+  }
+
```

Comment:
Both `capacity` and `max_size` would suggest that `async_mdarray` could be resized after construction.  The current design does not (or should not) permit post-construction resizing, since there are no constraints prohibiting resizing if all the extents are known at compile time.  Therefore, should we consider removing these functions?

```suggestion
```

Reply:
dropped